/*
proto
*/

syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.helloworld";
option java_outer_classname = "HelloWorldProto";
option objc_class_prefix = "HLW";

package network;

// Contains timestamp message type
import "google/protobuf/timestamp.proto";

service SNS{
    rpc InitConnect (ClientConnect) returns (ServerAllow) {}
    rpc Follow (FollowRequest) returns (FollowReply) {}
    rpc Unfollow (UnfollowRequest) returns (UnfollowReply) {}
    rpc List (ListRequest) returns (ListReply) {}
    rpc Update (UpdateRequest) returns (UpdateReply) {}
    rpc SendPost (Post) returns (PostReply) {}
    rpc ExecDebug (DebugRequest) returns (DebugReply) {}
} //end SNS

message ClientConnect {
    string connectingClient = 1;
}

message ServerAllow {
    int32 IReplyValue = 1;
}

message FollowRequest{
    // Client specifies name to follow
    string requestingClient = 1;
    string followRequest = 2;
}

message FollowReply{
    int32 IReplyValue = 1;
}

message UnfollowRequest{
    // Client specifies name to unfollow
    string requestingClient = 1;
    string unfollowRequest = 2;
}

message UnfollowReply{
    int32 IReplyValue = 1;
}

message ListRequest{
    // This can be empty
    string username = 1;
}

message ListReply{
    int32 IReplyValue = 1;
    repeated string users = 2;
    repeated string followers = 3;
}

message UpdateRequest {
    string username = 1;
}

message UpdateReply {
    Timeline updated = 1;
}

message PostReply {
    int32 IReplyValue = 1;
}

// Timeline View / Persistent Storage Message Types
message Post {
    string name = 1;
    google.protobuf.Timestamp time = 2;
    string content = 3;
}

message Timeline {
    repeated Post posts = 1;
}

message UserData {
    string name = 1;
    Timeline user_timeline = 2;
    repeated string following = 3;
    repeated string followers = 4;
}

message UserDataBase {
    map<string, UserData> all_users = 1;
}

// Debug
message DebugRequest {
    google.protobuf.Timestamp timein = 1;
}

message DebugReply {
    google.protobuf.Timestamp timeout = 1;
}
